scripts/runnable_recipe_ray_vllm_weight_sync.py

```diff
@@ -368,6 +369,447 @@ def run(self):
+    text_response_key: str = "text_response",
+    log_prob_key: str = "log_probs",
+    attention_mask_key: str = "attention_mask",
+    padding_value: int = -1,
+) -> Trajectory:
+    
+    args = ()
+    
+    if generate_kwargs is None:
+        generate_kwargs = {}
+    generate_kwargs.setdefault("detokenize", True)
+    generate_kwargs.setdefault("prompt_logprobs", False)
+    generate_kwargs.setdefault("logprobs", True)
+    # Create SamplingParams from generate_kwargs
+    sampling_params = SamplingParams(**generate_kwargs)
+    kwargs = {"sampling_params": sampling_params, "use_tqdm": False}
+
+    
+    txt = tensordict.get(text_key)
+    if not isinstance(txt, (list, str)):
+        txt = txt.tolist()
+    args = (txt,)
+
+    tokens_out = vllm_instance.generate(*args, **kwargs)
+    tokens_out = _get_output_tokens_and_log_probs(tokens_out, tokenizer, log_prob_key, token_response_key, text_response_key, padding_value)
+
+    in_keys = [
+        log_prob_key,
+        token_response_key,
+        text_response_key,
+        token_key,
+        attention_mask_key,
+    ]
+    out = tokens_out.select(*in_keys, strict=False)
+
+    # ================================================================
+    out[token_key] = tensordict[token_key]
+    # ================================================================
+
+    inputs = tensordict.select(text_key, strict=False)
+    if inputs.ndim < out.ndim:
+        inputs = inputs.unsqueeze(-1).expand(out.shape)
+    out.update(inputs)
+
+    td = tensordict.clone()
+    td.update(out, keys_to_update=list(out.keys()))
+
+    # This runs into errors with accessing resut["bla"] e.g. llm.py:511
+    # return Trajectory.from_tensordict(out)
+    return td 
+
+
+CompletionOutput_tc = from_dataclass(vllm.outputs.CompletionOutput)
+
+class _RequestOutput_tc(TensorClass["nocast"]):
+    request_id: str
+    prompt: str
+    prompt_token_ids: str
+    prompt_logprobs: str
+    outputs: str
+    finished: str
+    metrics: str
+    lora_request: str
+    encoder_prompt: str
+    encoder_prompt_token_ids: str
+    num_cached_tokens: str
+
+    def __post_init__(self):
+
+        def postproc(output):
+            def get_logprob(output):
+                t = []
+                for v, tid in zip(output.logprobs, output.token_ids):
+                    t.append(
+                        v[tid]["logprob"] if v[tid].get("logprob") is not None else 1.0
+                    )
+                return torch.tensor(t)
+
+            if output.logprobs:
+                output.logprobs = get_logprob(output)
+            output.token_ids = torch.tensor(output.token_ids)
+            return output
+
+        if isinstance(self.outputs, list):
+            outputs = self.outputs
+            outputs = [
+                postproc(from_dataclass(output, dest_cls=CompletionOutput_tc))
+                for output in outputs
+            ]
+            if len(outputs) == 1:
+                self.outputs = outputs[0]
```

Comment:
Why change our signature? A list of 1 element is perfectly fine

Reply:
I'm not following how this changes the signature

In the case where len(outputs) > 1 we do maybe_dense_stack so in both cases we do not return a list.